src/common/lua/ngx_wasm_lua_ffi.c

```diff
@@ -304,126 +288,126 @@ ngx_wa_ffi_shm_setup_zones(ngx_wa_ffi_shm_setup_zones_handler_pt setup_handler)
 
     if (node->right != sentinel) {
-        shm_kv_retrieve_keys(node->right, sentinel, start, limit, keys, total);
+        retrieve_keys(node->right, sentinel, limit, start_idx, cur_idx, keys);
     }
 }
 
 
 ngx_int_t
-ngx_wa_ffi_shm_iterate_keys(ngx_wa_shm_t *shm,
-                            ngx_uint_t page, ngx_uint_t page_size,
-                            ngx_str_t **keys, ngx_uint_t *total)
+ngx_wa_ffi_shm_iterate_keys(ngx_wa_shm_t *shm, ngx_uint_t page_size,
+    ngx_uint_t *clast_idx, ngx_uint_t *cur_idx, ngx_str_t **keys)
 {
+    ngx_uint_t        last_idx = clast_idx ? *clast_idx : 0;
     ngx_wa_shm_kv_t  *kv;
 
-    ngx_wa_assert(shm && shm->type != NGX_WA_SHM_TYPE_QUEUE);
+    ngx_wa_assert(shm->type != NGX_WA_SHM_TYPE_QUEUE);
 
     if (!ngx_wa_shm_locked(shm)) {
         return NGX_ABORT;
     }
 
     kv = shm->data;
 
-    if (page >= kv->nelts) {
-        return NGX_DECLINED;
+    if (last_idx >= kv->nelts) {
+        /* finished */
+        return NGX_DONE;
     }
 
-    if (kv->rbtree.root != kv->rbtree.sentinel) {
-        shm_kv_retrieve_keys(kv->rbtree.root, kv->rbtree.sentinel,
-                             page, page_size, keys, total);
-    }
+    /* cannot be empty here (kv->nelts > 0) */
+    ngx_wa_assert(kv->rbtree.root != kv->rbtree.sentinel);
 
-    *total = *total - page;
+    retrieve_keys(kv->rbtree.root, kv->rbtree.sentinel, page_size, last_idx,
+                  cur_idx, keys);
+
+    *cur_idx -= last_idx;
+
+    if (clast_idx) {
+        *clast_idx += *cur_idx;
+    }
 
     return NGX_OK;
 }
 
 
+ngx_uint_t
+ngx_wa_ffi_shm_kv_nelts(ngx_wa_shm_t *shm)
+{
+    ngx_wa_shm_kv_t  *kv;
+
+    ngx_wa_assert(shm->type != NGX_WA_SHM_TYPE_QUEUE);
+
+    kv = shm->data;
+
+    return kv->nelts;
+}
+
+
 ngx_int_t
-ngx_wa_ffi_shm_get_kv_value(ngx_wa_shm_t *shm,
-                            ngx_str_t *k, ngx_str_t **v, uint32_t *cas)
+ngx_wa_ffi_shm_kv_get(ngx_wa_shm_t *shm, ngx_str_t *k, ngx_str_t **v,
+    uint32_t *cas)
 {
-    unsigned   unlock = 0;
     ngx_int_t  rc;
 
-    ngx_wa_assert(shm && k && v && cas);
+    ngx_wa_assert(shm->type == NGX_WA_SHM_TYPE_KV);
 
-    if (!ngx_wa_shm_locked(shm)) {
-        ngx_wa_shm_lock(shm);
-        unlock = 1;
-    }
+    ngx_wa_shm_lock(shm);
 
     rc = ngx_wa_shm_kv_get_locked(shm, k, NULL, v, cas);
 
-    if (unlock) {
-        ngx_wa_shm_unlock(shm);
-    }
+    ngx_wa_shm_unlock(shm);
 
     return rc;
 }
 
 
 ngx_int_t
-ngx_wa_ffi_shm_set_kv_value(ngx_wa_shm_t *shm,
-                            ngx_str_t *k, ngx_str_t *v, uint32_t cas,
-                            ngx_int_t *written)
+ngx_wa_ffi_shm_kv_set(ngx_wa_shm_t *shm, ngx_str_t *k, ngx_str_t *v,
+    uint32_t cas, ngx_int_t *written)
 {
-    unsigned   unlock = 0;
     ngx_int_t  rc;
 
-    ngx_wa_assert(shm && k && v && written);
+    ngx_wa_assert(shm->type == NGX_WA_SHM_TYPE_KV);
 
-    if (!ngx_wa_shm_locked(shm)) {
-        ngx_wa_shm_lock(shm);
-        unlock = 1;
-    }
+    ngx_wa_shm_lock(shm);
```

Comment:
Ditto not allowing the user to set values while iterating over the keys of an shm.

Reply:
Is it really a good idea to do that? If so there should be tests for this, at the minimum.